src/http/modules/ngx_http_proxy_module.c

```diff
@@ -2847,6 +2741,57 @@ ngx_http_proxy_add_x_forwarded_for_variable(ngx_http_request_t *r,
 }
 
 
+static ngx_int_t
+ngx_http_proxy_internal_host_variable(ngx_http_request_t *r,
+    ngx_http_variable_value_t *v, uintptr_t data)
+{
+    ngx_http_proxy_ctx_t  *ctx;
+
+    ctx = ngx_http_get_module_ctx(r, ngx_http_proxy_module);
+
+    if (ctx == NULL || ctx->authority_set) {
+        v->not_found = 1;
+        return NGX_OK;
+    }
+
+    v->len = ctx->vars.host_header.len;
+    v->valid = 1;
+    v->no_cacheable = 0;
+    v->not_found = 0;
+    v->data = ctx->vars.host_header.data;
+
+    return NGX_OK;
+}
+
+
+static ngx_int_t
+ngx_http_proxy_internal_connection_variable(ngx_http_request_t *r,
+    ngx_http_variable_value_t *v, uintptr_t data)
+{
+    ngx_http_proxy_ctx_t  *ctx;
+
+    ctx = ngx_http_get_module_ctx(r, ngx_http_proxy_module);
+
+    if (ctx == NULL || ctx->connection_type == 0) {
+        v->not_found = 1;
+        return NGX_OK;
+    }
+
+    v->valid = 1;
+    v->no_cacheable = 0;
+    v->not_found = 0;
+
+    if (ctx->connection_type == NGX_HTTP_CONNECTION_CLOSE) {
+        ngx_str_set(v, "close");
+
+    } else if (ctx->connection_type == NGX_HTTP_CONNECTION_KEEP_ALIVE) {
+        ngx_str_set(v, "keep-alive");
+    }
```

Comment:
The newly added field `connection_type` is only set hardcoded to NGX_HTTP_CONNECTION_CLOSE
in ngx_http_proxy_module.c, so checking for NGX_HTTP_CONNECTION_KEEP_ALIVE looks unused.

Further, adding this field can be avoided at all if replaced with a check based on the HTTP version.
This could be done by using http_version from proxy configuration; accessing configuration from
a variable handler should be good, we do similar things e.g. in ngx_http_variable_sent_keep_alive()).

Another option is to reuse `authority_set` (with renaming to some more generic name like "legacy_set").
The latter would make it look similar to how we check `authority_set` in ngx_http_proxy_internal_host_variable().
```suggestion
    if (ctx == NULL || ctx->legacy_set == 0) {
        v->not_found = 1;
        return NGX_OK;
    }

    v->valid = 1;
    v->no_cacheable = 0;
    v->not_found = 0;
    ngx_str_set(v, "close");
```

Reply:
Maybe just `ctx->legacy`.

Reply:
I used `legacy`, and let me know if there is a better one.